Source/wayland/WaylandServer.m

```diff
@@ -0,0 +1,1482 @@
+	window = wlconfig->pointer.focus;
+	NSEvent *event;
+	NSEventType eventType;
+	NSPoint eventLocation;
+	NSGraphicsContext *gcontext;
+	unsigned int eventFlags;
+	float deltaX = sx - window->wlconfig->pointer.x;
+	float deltaY = sy - window->wlconfig->pointer.y;
+
+	NSDebugLog(@"obtaining locations: wayland=%fx%f pointer=%fx%f",
+		   sx, sy, window->wlconfig->pointer.x, window->wlconfig->pointer.y);
+
+	gcontext = GSCurrentContext();
+	eventLocation = NSMakePoint(sx,
+				    window->height - sy);
+
+	eventFlags = 0;
+	eventType = NSLeftMouseDragged;
+
+	NSDebugLog(@"sending pointer delta: %fx%f, window=%d", deltaX, deltaY, window->window_id);
+
+	event = [NSEvent mouseEventWithType: eventType
+				   location: eventLocation
+			      modifierFlags: eventFlags
+				  timestamp: (NSTimeInterval) time / 1000.0
+			       windowNumber: (int)window->window_id
+				    context: gcontext
+				eventNumber: time
+				 clickCount: 1
+				   pressure: 1.0
+			       buttonNumber: 0 /* FIXME */
+				     deltaX: deltaX
+				     deltaY: deltaY
+				     deltaZ: 0.];
+
+	[GSCurrentServer() postEvent: event atStart: NO];
+    }
+
+    wlconfig->pointer.x = sx;
+    wlconfig->pointer.y = sy;
+}
+
+static void
+pointer_handle_button(void *data, struct wl_pointer *pointer, uint32_t serial,
+		      uint32_t time, uint32_t button, uint32_t state_w)
+{
+    NSDebugLog(@"pointer_handle_button: button=%d", button);
+    WaylandConfig *wlconfig = data;
+    NSEvent *event;
+    NSEventType eventType;
+    NSPoint eventLocation;
+    NSGraphicsContext *gcontext;
+    unsigned int eventFlags;
+    float deltaX = 0.0;
+    float deltaY = 0.0;
+    int clickCount = 1;
+    int tick;
+    enum wl_pointer_button_state state = state_w;
+    struct window *window = wlconfig->pointer.focus;
+
+    gcontext = GSCurrentContext();
+    eventLocation = NSMakePoint(wlconfig->pointer.x,
+				window->height - wlconfig->pointer.y);
+    eventFlags = 0;
+    if (state == WL_POINTER_BUTTON_STATE_PRESSED) {
+	if (button == wlconfig->pointer.last_click_button &&
+	    time - wlconfig->pointer.last_click_time < 300 &&
+	    abs(wlconfig->pointer.x - wlconfig->pointer.last_click_x) < 3 &&
+	    abs(wlconfig->pointer.y - wlconfig->pointer.last_click_y) < 3) {
+	    NSDebugLog(@"handle_button HIT: b=%d t=%d x=%f y=%f", button, time, wlconfig->pointer.x, wlconfig->pointer.y);
+	    wlconfig->pointer.last_click_time = 0;
+	    clickCount++;
+	} else {
+	    NSDebugLog(@"handle_button MISS: b=%d t=%d x=%f y=%f", button, time, wlconfig->pointer.x, wlconfig->pointer.y);
+	    wlconfig->pointer.last_click_button = button;
+	    wlconfig->pointer.last_click_time = time;
+	    wlconfig->pointer.last_click_x = wlconfig->pointer.x;
+	    wlconfig->pointer.last_click_y = wlconfig->pointer.y;
+	}
+
+	switch (button) {
+	case BTN_LEFT:
+	    eventType = NSLeftMouseDown;
+	    break;
+	case BTN_RIGHT:
+	    eventType = NSRightMouseDown;
+	    break;
```

Comment:
What happens for all these other buttons on my mouse?

Reply:
The third constants I found supported is `NSOtherMouse{Down,Up}`, so I added support for that.

If other buttons on your mouse are supported by libinput, I don't immediately know how to map them to AppKit constants: `BTN_SIDE`, `BTN_EXTRA`, `BTN_FORWARD`, `BTN_BACK` and any other ones. :-)